### ISSUE TITLE
Add funtion calling (CRUD) on a database

### ISSUE THREAD

**Issue by Rauljauregi** [ragapp/ragapp#68]
**Is your feature request related to a problem? Please describe.**
Expand RagApp capabilities by integrating additional functionalities. Specifically, the ability to manage function data and perform CRUD operations from the admin panel could enable the creation of more advanced AI agents.

**Describe the solution you'd like**
I propose adding a function calling feature to the admin panel. This feature would allow administrators to upload function data in JSON format, adhering to the OpenAI style. Additionally, it would enable CRUD (Create, Read, Update, Delete) operations on the database. This could potentially be integrated with our existing vector database, though we may need to consider using a separate SQL database for these operations.

**Describe alternatives you've considered**
- Continuing with the current focus on RAG applications only, which limits the scope of what we can achieve.

**Additional context**
Adding this feature would extend the RAGAPP project's capabilities, allowing us to develop more sophisticated AI agents. Key points to consider include:
- Ensuring JSON schema validation for function data uploads to maintain data integrity.
- Providing a user-friendly interface for CRUD operations within the admin panel.
- Assessing the feasibility and performance implications of integrating these operations with the existing vector database versus using a separate SQL database.

By implementing this feature, we could significantly enhance the flexibility and functionality of the RAGAPP project, moving beyond its current scope to support a broader range of AI applications.

**Comment by marcusschiesser** [comment]
@Rauljauregi how about wrapping your CRUD operations into an OpenAPI and adding it using the new OpenAPI actions feature?

**Comment by Rauljauregi** [comment]
Yes it would be great, do you want me to remove the feature suggestion? Or to edit it?

Thank you very much, great project RagAPP :)

**Comment by marcusschiesser** [comment]
Thanks @Rauljauregi - yes please give OpenAPI a try and update this feature request accordingly

**Comment by Rauljauregi** [comment]
I tried my best but with no success. On other project I have, I managed to do function calling, but I use, OpenAi, funcion feature, (CRUD json data schema, and the prompt refering to the functions and its properties) and  Botpress to do Webhook to the database CRUD for those functions.


I think this is not possible with OpenAPI, or at least i am not able to do that.

**Comment by marcusschiesser** [comment]
how does your OpenAPI spec look like?

**Comment by Rauljauregi** [comment]
I ha e tried many things. This is the last one, with a local PostgreSQL local database. https://vialabsdigital.com/wp-content/uploads/openapi_spec_postgresql.yaml

**Comment by Rauljauregi** [comment]
Please evaluate the following suggestion for making this feature possible. It might not be the best approach, but I hope you find it interesting:

Proposal:

### Admin Options:

Add choices in the admin area to select the provider (e.g., OpenAI, Anthropic) and the specific agent.
In the case of OpenAI, this would allow the use of the function calling feature.

### Function Calling:

Enable the system to listen for function calls from the selected agents.
Not sure how to set up the ability to do webhooks to perform CRUD operations on a connected database, or if this can be fully achieved using OpenAI alone.
Thanks,

**Comment by marcusschiesser** [comment]
@Rauljauregi I think it would be valuable to add a docker compose file that adds ragapp together with a postgres.

Your OpenAPI spec:  https://vialabsdigital.com/wp-content/uploads/openapi_spec_postgresql.yaml must then be modified to use the URL of the DB inside that docker network - this should work

Also you should add a description to the OpenAPI spec - that's needed for the LLM to know whether to call the tool or not

**Comment by marcusschiesser** [comment]
@Rauljauregi want to have a look into this?

**Comment by Rauljauregi** [comment]
Thank you! 😀👏🏻
